lib/Sema/TypeCheckType.cpp

```diff
@@ -2846,6 +2846,19 @@ TypeResolver::resolveAttributedType(TypeAttributes &attrs, TypeRepr *repr,
   }
 
   if (hasFunctionAttr && !fnRepr) {
+
+    if (attrs.has(TAK_Sendable)) {
+      // A Sendable closure must be attached to a function type. If the
+      // attribute is attached to a typealias that points to a function type,
+      // then we won't have a FunctionTypeRepr. So just ignore the repr and just
+      // check whether we have a function type. If we don't, the code below will
+      // emit a diagnostic later on.
+      if (ty->is<FunctionType>()) {
+        // Ok, remove the attribute from the set so we don't diagnose.
+        attrs.clearAttribute(TAK_Sendable);
```

Comment:
Oh sorry, I think I misunderstood the `attrs` variable (I thought we were working with a copy of the attributes for diagnostic purposes and not removing tha attributes directly from the declaration). I’ll tweak it so we don’t drop it.

Reply:
Yeah, `resolveAttributedType` gets a reference to `repr` attributes, so dropping an attribute means that repr is going to loose it too.

Reply:
@theblixguy ^

Reply:
Oh haha no worries! I will add a test case :)

Reply:
Actually, the parameter and function types don't need to conform to Sendable. I was a bit confused why we were not emitting diagnostics for it so I looked into it and found this change by Doug who removed the diagnostic based on pitch feedback - https://github.com/apple/swift/pull/35915/commits/5af310a5c4e9a8deb79e0bb6d97aec1256cc031c

Reply:
So what does it mean to have Sendable attribute attached to the function?

Reply:
For context see this - https://forums.swift.org/t/pitch-4-concurrentvalue-and-concurrent-closures-evolution-pitches/44446/14

Also BTW we do seem to be dropping `@Sendable` but it's because we call `resolveASTFunctionType` in the `hasFunctionAttr && fnRepr` branch above (which sets the `Sendable` bit on the type if it's in the `attrs` list):

https://github.com/apple/swift/blob/9c4972ff8470b50fc01716278947311ef6bd8b7a/lib/Sema/TypeCheckType.cpp#L2724-L2728

We are not setting that bit, so my test case is failing:

```swift
typealias VoidFunction = () -> Void

func a1(callback: @Sendable VoidFunction) {}
func a2(callback: @Sendable () -> Void) {}

func b() {
  let nonSendableFn: () -> Void = {}
  a1(callback: nonSendableFn) // no warning, but should warn
  a2(callback: nonSendableFn) // warning
}
```

I don't think I can call `resolveASTFunctionType` outside that branch because I don't have a `fnRepr`, but perhaps I can set tweak the type's `ExtInfo` in my branch? That's basically what `resolveASTFunctionType` is doing inside. WDYT?

Reply:
I gave it a try and it works (I tried it with various test cases). Basically we just need to set the `Sendable` bit in the `ExtInfo` of the function type, which is what we're doing anyway but we just don't do it when we have a typealias.

Reply:
That sounds reasonable to me! /cc @kavon and @DougGregor

Reply:
Thank you! I have pushed my changes.